### DISCUSSION TITLE
Coroutine does not work asyncronously

### DISCUSSION THREAD

**Discussion by ghost** [reactphp/reactphp#553] [Q&A]
I have this code for my discord bot. I am using DiscordPHP library that is built on ReactPHP.

```php
// SubscriptionBot.php 

// ...

$this->discord->getLoop()->addPeriodicTimer(10, function () {
    echo 'a';
    $expired_subs = Database::getExpiredSubscriptions();
    if (!empty($expired_subs))
        foreach ($expired_subs as $expired_sub) $this->suspendSubscription($expired_sub);
    echo 'c';
});

$this->discord->getLoop()->addTimer(10.1, function () {
    echo 'b';
});

// ...
``` 

```php
private function suspendSubscription($subscription): void
{
    coroutine(function () use ($subscription) {
        Database::deleteSubscription($subscription['id']);
        $guild = yield $this->discord->guilds->fetch($subscription['guild_id']);
        $member = yield $guild->members->fetch($subscription['user_id']);
        $member->removeRole($_ENV['DISCORD_ROLE_ID']);
        $channel = yield $guild->channels->fetch($_ENV['DISCORD_NOTIFY_CHANNEL_ID']);
        $channel->sendMessage($member . "**, your subscription has expired.** Use **/purchase** to renew.");
    });
}
```

This outputs: `acb`

If I change code and use async()/await(), everything works just fine: `abc`

Maybe I misunderstood coroutine's purpose?

**Comment by WyriHaximus** [comment]
Both coroutine and async/await have the goal of making using promises a lot easier to read. They differ in how they work behind the scenes but you should see the same result. Given you seen to be dealing with a database and the Discord API that 0.1 second could trigger both earlier or later then it completes your coroutine.

**Comment by yadaiio** [comment]
Hi @sherwomeister, thank you for bringing this up. Could you please confirm if you are using PHP 8.1+? If you are, I would recommend using [async v4](https://github.com/reactphp/async) and await with fibers for better performance.

It seems like you are attempting to execute commands in a specific order, but encountering timing issues. Depending only on timers can be unpredictable, unless it is for demonstration purposes in your example. I suggest considering the use of await for each individual time step instead.
